samples: coremark: Add coremark sample to sdk-nrf (#14315)
nrfconnect/sdk-nrf: doc/nrf/external_comp/coremark.rst

> .. _integration_template:

Reviewer: duplicate label error
```suggestion
.. _ug_coremark:
```

CoreMark integration
####################

.. contents::
   :local:
   :depth: 2

> CoreMark is a CPU benchmark provided by EEMBC (Embedded Microprocessor Benchmark Consortium) to verify SoCs performance.
> It provides a series of test done in each benchmark iteration.
> The number of iterations DUT can perform per second gives benchmark score.
> Integrating CoreMark into |NCS| allows users to easily verify performance of acquired SoCs.

Reviewer: ```suggestion
CoreMark is a CPU benchmark provided by Embedded Microprocessor Benchmark Consortium (EEMBC) to verify SoCs performance.
You can use it to perform series of tests in each benchmark iteration.
The benchmark score is calculated based on the number of iterations the device under test (DUT) can perform.
Integrating CoreMark into the |NCS| allows users to easily verify SoCs performance.
```

Integration prerequisites
*************************

Before you start the |NCS| integration with CoreMark, make sure that the following prerequisites are completed:

* :ref:`Installation of the nRF Connect SDK <installation>`.
* Setup of the required :term:`Development Kit (DK)`.

Solution architecture
*********************

> CoreMark repository is added by default as a module to |NCS| by adding CoreMark repository to |NCS| west manifest.

Reviewer: ```suggestion
By adding the CoreMarek repository to the |NCS| west manifest, you can include it in the repository as the |NCS| module.
```

Author: User doesn't have to do anything he already has coremark repository thanks to west manifest.

Reviewer: Edited the above suggestion - please double-check

> It can be found under path :file:`modules/benchmark/coremark`.

Reviewer: ```suggestion
You can find it under the :file:`modules/benchmark/coremark` path.
```

> To run CoreMark correctly SoC needs a porting layer.
> The porting layer for devices supported in |NCS| was added and placed in: :file:`nrf/modules/coremark`.

Reviewer: ```suggestion
CoreMark needs a porting layer to run correctly on an SoC.
The porting layer for devices supported in the |NCS| are under the :file:`nrf/modules/coremark` path.
```

Author: "an SoC" ?  Shouldn't it be "a SoC"?

Reviewer: I think we've had a similar discussion before with the team regarding the HID report. 
Similarly to that case, it depends on how you read it (each letter separately, starting with 'eS', and that puts the vowel first, or as a one word, in which case the first letter is a consonant). 
So both are correct but we can definitely keep "a" as I believe it's generally more common, so let's leave as is.

CoreMark benchmark consists of continuously repeated iterations.

> Each iteration in turn consists of workloads: Linked List, Matrix Multiply, and State Machine.
> For detailed description of each of those workloads please reefer to CoreMark README: `CoreMark GitHub`.

Reviewer: ```suggestion
Each iteration, in turn, consists of the following workloads: linked list, matrix multiply, and state machine.
For detailed description of each of those workloads, see the `CoreMark GitHub`_ repository.
```

Integration steps
*****************

> To integrate CoreMark into |NCS|, after completing the prerequisites, complete the following steps:

Reviewer: ```suggestion
To integrate CoreMark into the |NCS|, complete the following steps:
```
Since they're prerequisites it's probably not necessary to repeat that.
Not really sure if this description is enough overall (e.g., template refers to this as an example: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/external_comp/bt_fast_pair.html#ug-integrating-fast-pair )

Author: This integration is much simpler then fast pair so it doesn't have such big  Integration section

Configuring CoreMark
====================

To include CoreMark in your build, add the following Kconfig options in your :file:`prj.conf` file:

.. code-block:: console

   CONFIG_COREMARK=y

> 
> As a next step choose memory method for dealing with CoreMark data.
> Make sure that you allocate enough memory depending on the method you choose.
> 
> 1. Memory method stack
> 
> .. code-block:: console
> 
>    COREMARK_MEMORY_METHOD_STACK=y
>    CONFIG_MAIN_STACK_SIZE=4096
> #. Memory method heap
> 
> .. code-block:: console
> 
>    CONFIG_COREMARK_MEMORY_METHOD_MALLOC=y
>    CONFIG_HEAP_MEM_POOL_SIZE=4096
> 
> #. Memory method static
> 
> .. code-block:: console
> 
>    CONFIG_COREMARK_MEMORY_METHOD_STATIC=y

Reviewer: ```suggestion

#. Choose the memory method for dealing with the CoreMark data.
   Make sure that you allocate enough memory depending on the method you choose.

   * Memory method stack

   .. code-block:: console

      COREMARK_MEMORY_METHOD_STACK=y
      CONFIG_MAIN_STACK_SIZE=4096

   * Memory method heap

    .. code-block:: console

      CONFIG_COREMARK_MEMORY_METHOD_MALLOC=y
      CONFIG_HEAP_MEM_POOL_SIZE=4096

   * Memory method static

    .. code-block:: console

      CONFIG_COREMARK_MEMORY_METHOD_STATIC=y
```

> 
> For an example, see the :ref:`_coremark_configuration` section of the :ref:`coremark_sample` documentation.

Reviewer: ```suggestion
To see an example, refer to the :ref:`coremark_configuration` section of the :ref:`coremark_sample` documentation.
```

> 

Reviewer: ```suggestion
```

Submitting score
================

> To see scores already submitted to CoreMark see official database at `CoreMark®`.
> If you want to submit your own score please follow instructions found at official CoreMark repository README: `CoreMark GitHub`.

Reviewer: ```suggestion
To see the scores already submitted to CoreMark, see the official database at `CoreMark®`_.
If you want to submit your own score, follow steps in the `CoreMark GitHub`_ repository.
```

Applications and samples
************************

> The following samples demonstrate the CoreMark integration in |NCS|:

Reviewer: ```suggestion
The following samples demonstrate the CoreMark integration in the |NCS|:
```

> * :ref:`coremark_sample` - allows to get CoreMark score for boards supported in |NCS|.

Reviewer: ```suggestion
* :ref:`coremark_sample` - This sample is used to obtain the CoreMark score for boards supported in the |NCS|.
```

.. _ug_coremark:

CoreMark integration
####################

.. contents::
   :local:
   :depth: 2

> 

Reviewer: ```suggestion
```

> CoreMark is a CPU benchmark provided by Embedded Microprocessor Benchmark Consortium (EEMBC) to verify SoCs performance.

Reviewer 2: ```suggestion
CoreMark® is a CPU benchmark provided by Embedded Microprocessor Benchmark Consortium (EEMBC) to verify SoCs performance.
```

Before you start the |NCS| integration with CoreMark, make sure that the following prerequisites are completed:

* :ref:`Installation of the nRF Connect SDK <installation>`.
* Setup of the required :term:`Development Kit (DK)`.

Solution architecture
*********************

Adding the CoreMark repository to the |NCS| manifest automatically includes it in the repository as the |NCS| module.
You can find it under the :file:`modules/benchmark/coremark` path.
CoreMark needs a porting layer to run correctly on a SoC.
The porting layer for devices supported in the |NCS| are under the :file:`nrf/modules/coremark` path.

CoreMark benchmark consists of continuously repeated iterations.
Each iteration, in turn, consists of the following workloads: linked list, matrix multiply, and state machine.
For detailed description of each of those workloads, see the `CoreMark GitHub`_ repository.

Integration steps
*****************

To integrate CoreMark into the |NCS|, complete the following steps:

Configuring CoreMark
====================

To include CoreMark in your build, add the following Kconfig options in your :file:`prj.conf` file:

.. code-block:: console

   CONFIG_COREMARK=y

> #. Choose the memory method for dealing with the CoreMark data.

Reviewer 2: No numbered list that I can see, so remove #.

   Make sure that you allocate enough memory depending on the method you choose.

   * Memory method stack

   .. code-block:: console

      COREMARK_MEMORY_METHOD_STACK=y
      CONFIG_MAIN_STACK_SIZE=4096

   * Memory method heap

    .. code-block:: console

      CONFIG_COREMARK_MEMORY_METHOD_MALLOC=y
      CONFIG_HEAP_MEM_POOL_SIZE=4096

   * Memory method static

> 
> .. code-block:: console
> .. code-block:: console

Reviewer: ```suggestion
```

    .. code-block:: console

      CONFIG_COREMARK_MEMORY_METHOD_STATIC=y

To see an example, refer to the :ref:`coremark_configuration` section of the :ref:`coremark_sample` documentation.

Submitting score
================

> To see the scores already submitted to CoreMark, see the official database at `CoreMark®`_.

Reviewer: ```suggestion
You can see the submitted scores in the official `CoreMark®`_ database.
```

.. _ug_coremark:

CoreMark integration
####################

.. contents::
   :local:
   :depth: 2

CoreMark® is a CPU benchmark provided by Embedded Microprocessor Benchmark Consortium (EEMBC) to verify SoCs performance.
You can use it to perform series of tests in each benchmark iteration.

> The benchmark score is calculated based on the number of iterations per second the device under test (DUT) can perform.

Reviewer 3: ```suggestion
The benchmark score is calculated based on the number of iterations per second the :term:`Device Under Test (DUT)` can perform.
```
